### ISSUE TITLE
Permission to use, learn from that code?

### ISSUE THREAD

**Issue by TTAJ1b14** [boltzer/BZ_LCJ80#1]
@boltzer 

Hi! Wonder to get your thoughts on your creation permissions. I am not professional coder, just amateur, who is ready to click and try through all options befoe something finally starts to work. 

I see well-written code here that can that can teach an amateur like me a lot. And I am the guy interested in creation of new vehicles in free time.

My question if I can get your permission to play with this code, modify and use it? And if I can get it, I have couple more questions regarding logic implementation if you don't mind :)

Thank you in advance for your answer

**Comment by boltzer** [comment]
You are welcome to use this code for learning purposes, including modification. It's a good way to learn the ropes, I just ask that you don't repost the mod to the workshop as your own.

Happy to answer any questions you have, but forgive me if I can't respond immediately

**Comment by TTAJ1b14** [comment]
Appreciate a lot! 

Defenitely not gonna repost. Ideas are to learn, improve current works and see if I can implement armor logic I wanted fo long time with proper data handling.

Briefly speaking, working on how to create separate distinct types of armor, depending on item installed in part slot. Might interesting idea for your future features as well

Let's say wooden planks on doors that give small protection for windows and doors is type 1 item and armor
Metal sheets giving much bigger protection installed at the same place is type 2 armor and item.

Thanks one more time and have a great day!

**Comment by boltzer** [comment]
best way to distinguish armor capability between types of armors would be in the lua-side of things, there's no vanilla implementation of armor AFAIK. 

The way that my implementation works is by monitoring the condition of the "protected" parts, for example the hood. Once it detects that the hood has taken damage and the condition has gone down, lets say from 100% to 80%, it routes that to the front bumper. But there are two main conditions in the case that the hood receives 20% damage:

1) Front bumper is >= 66% health: Bumper absorbs ALL of the damage(determined by `absorption,` and receives a fraction of the damage determined by `factor`, which for standard bumper is .85. Hood receives no damage, bumper receives (20 * 0.85) = 17% damage.
2) Front bumper is < 66% health: Bumper absorbs 20% of the damage, so hood still gets 80% of the damage. bumper still receives 85% of the damage. Hood takes (20 * 0.8) = 16% damage, bumper takes (20 * 0.85) = 17% damage.

For an armor made of wood planks you would just need to define the recipe for that, otherwise implementation would be the same as above. For me, factor and absorption variables is how I differentiate between armor tiers.

**Comment by TTAJ1b14** [comment]
Yep, you are right. The vanilla game is missing any armor logic, except engine protection.

Appreciate your explanation of the logic! Couple questions if you don't mind :)

1) Does your system allows to create self-protecting parts? Let's say an armored vehicle will have door that initially should take more damage then a door of any other vehicle? If I assign door as a part that should be protected and door as a part that is protecting. Will it result in damage self absortion? Or infinite health as a bug? Interesting point to try myself on weekends

2) If you don't mind, could you share a small note what is what here?

parts = {
        FrontBumper = {
            BZ_LCJ80FrontBumper = {
                FrontBumper0 = "BZ_LCJ80FrontBumper0",
                FrontBumper1 = "BZ_LCJ80FrontBumper1",
            },
            default = "first",
            noPartChance = 0,
        },

3. How system will handle anomaly sitiations like one part protecting 3 others and they get damage at the same time? Will protection part take triple damage or somehow divide it by 3? And vice versa. What if couple parts will protect one? Like window is protected by door and by door armor

Sorry for too many questions, but these is extremely interesting system you made and I can't help myself, but ask =)

**Comment by boltzer** [comment]
1. it would work a similar way, with some slight changes.

some variables to help explain:
`recordedCondition` is the condition that is kept track of in mod data. This is the source of truth for part condition and is only updated once all of this is said and done.
`currentCondition` is what is being checked basically every frame, where we ask the game what the part's condition is and compare it to recordedCondition.
`conditionDelta` is the difference between these two == (recordedCondition - currentCondition). 
-->currentCondition is ALWAYS less than or equal to recorded condition.
`newCondition` is what you want the condition to be corrected to.
it is always true that: 
currentCondition <= newCondition <= recordedCondition.

So you would need to monitor the door condition, and once a delta is detected, determine what fraction of the delta you want the door to take as damage.

two important things: 
1) newCondition is never higher than recordedCondition, because otherwise you introduce the possibility of a runaway infinite health bug
2) when you calculate the new condition, set the recorded condition in the mod data to prevent another infinite loop, only ending with 0 health instead of infinite health

2. this is basically how we represent the parts as a class/object. Lua isn't object-oriented, but you can trick it by using tables.

parts = a table representing the parts we want defined for use in lua
FrontBumper = a table representing the part, simplified name
BZ_LCJ80FrontBumper corresponds to the name of the front bumper part
BZ_LCJ80FrontBumper0 and BZ_LCJ80FrontBumper1 refers to the two variants of bumpers, corresponding to the name of their respective items. Note the distinction between a part and an item.

default means that the vehicle will always spawn the first variant---the second variant can only be crafted.
noPartChance = 0 means this part always spawns. 

3. Yes, I left this out of the original explanation for simplicity. You would just take into account how many parts an armor is protecting, so divide by 3.

you can also have multiple parts protect one part, you start by detecting damage taken(same way as above), and then route the damage to the parts that you want.

**Comment by boltzer** [comment]
@bknhrst This is a serious accusation that I take seriously and don't appreciate. I've unlisted the mod for the time being and I'd like to know what part of the code you specifically feel ownership over? As I mentioned in the mod discussion, the decisions I made with the implementation of this mod were informed by my time using vehicle mods from a variety of mod authors, KI5 included.

**Comment by boltzer** [comment]
because it is my own work? The model, textures, animations, and code. you seem to forget that there's a dearth of learning materials for vehicle modding in PZ, in which the only way to learn is to look at what's currently out there on the workshop. These code fragments you're referring to are left over from my learning process in which I iterate off of what I know works. Do I also have to remind you that, this is a video game mod, and we all work off the foundation that TIS built. I'm not being paid for any of this and have no intention of receiving money for it.

> i don't have the pull or reach to make you stop but you shouldn't be playing the innocent victim of a slander campaign here when all the evidence is out there and even somebody who doesn't know lua can compare the codes and come to the above mentioned conclusion.

Are you proud of yourself the way you talk to someone you don't know and make no effort to understand? If I wanted to play the victim, I would be going about this a VERY different way.

**Comment by boltzer** [comment]
And let me remind you, do DO have the pull and reach. I have no power over you within the PZ community, but you AND KI5 have power over me, and it's disingenuous to act otherwise.

**Comment by boltzer** [comment]
No one is making you escalate this but yourself. Your problem is that you feel that you have exclusive ownership over code that isn't even entirely yours to begin with. This misunderstanding stemmed from the fact that I, like many other modders, looked towards what was already on the PZ workshop and sought out to develop my own solution from that foundation. If it were as easy as you say to simply refactor code and rename variables, there would be a lot more people doing it.

I'm trying my best to keep it civil because, as I said, I have no incentive to plagiarize and steal your thunder, but it's clear there's nothing I can say to diffuse your frustration. You're an accomplished modder, and I still hold a great amount of respect for what you and KI5 have done for the community, but this interaction just reeks of a modder going on a power trip on someone who has NO influence over them.

**Comment by TTAJ1b14** [comment]
I'm not in a position to judge who's right or wrong in this debate, and I won't intervene in the discussion. However, out of fairness and respect, @bknhrst  , I consider it appropriate to ask here for your permission as well for the my main question stated at the very beginning of this discussion.

Thank you in advance for your response

**Comment by boltzer** [comment]
@TTAJ1b14 

If you haven't already I strongly recommend looking through the game files. They are, to me, the standard when it comes to getting an idea of how vehicles function and as a result, you'll find that all vehicle mods are some variation of this. I was going to mention this resource the other day but, well you know what happened. You'll find them under these directories:

C:\Program Files (x86)\Steam\steamapps\common\ProjectZomboid\media\scripts\vehicles\template_engine.txt
C:\Program Files (x86)\Steam\steamapps\common\ProjectZomboid\media\lua\server\Vehicles\Vehicles.lua

Vehicles.lua in particular answered a lot of questions I had when starting out, and I hope you find value from it as well. I'm sorry that you were caught in this, and I hope this doesn't discourage you from pursuing modding. I was happy to answer your questions because I was in your position when starting out a few months ago, and I understand that while the two tutorials most people will point you towards are good starting points, they don't really answer the questions of why, and some of the information is rather outdated. It can be frustrating, but also incredibly rewarding once you _do_ get things to work.

**Comment by TTAJ1b14** [comment]
Thank you one again for sharing your path!

Yep, already looked through this lines and  on moreover first vehicle is in workshop for some time. 

Works perfectly in Singleplayer. Currently the challenge is to make sure armor will be working properly im MP, which I have some doubts about. That's where all my questions start to arise about proper data handling and things that I obviously miss  when coming to MP part.

**Comment by TTAJ1b14** [comment]
@bknhrst,  

This is awesome library and thanks to its creator things like part-item based models became real for me, but armor logic implemented there goes a bit awry with what I am trying to implement.  Correct me if I am mistaken, but tsar's logic goes with the idea of complete part invincibility if armor part spotted installed. It's nice way, recognised by community, but not what I feel the armor should work.

My final goal is part-item based defence levels and calculations that will make vehicle stronger, but not invincible . However, currently my knowledge at this point is quite limited and that's why I came here with my questions.
